File: src/shared/engine/negamax.ts
import type { Chess } from "chessops/chess";

import { createDescend, type Descend, createDrawTest, legalMoves, type Repetition } from "../chess";
import type { PlayedMove, WeightVector } from "../eval";
import { createLeaf, type Leaf } from "./leaf";
import { orderMoves } from "./ordering";

// What a position the game cannot continue from is worth to the side standing in it. See
// `createDrawTest` for why it is a flat zero and not a tunable contempt.
export const DRAW_SCORE = 0;

const INFINITY = Number.POSITIVE_INFINITY;

// Everything one search holds for the whole of its life: where it puts the position it is trying,
// what it has already stood in, and what it does when it stops descending.
export type SearchContext = {
	descend: Descend;
	repetition: Repetition;
	drawn: (position: Chess) => boolean;
	leaf: Leaf;
};

export type Frame = {
	context: SearchContext;
	position: Chess;
	played?: PlayedMove;
	depth: number;
	ply: number;
	alpha: number;
	beta: number;
};

export function createSearchContext({
	weights,
	options,
	repetition,
}: {
	weights: WeightVector;
	options: { quiescence?: boolean; nodeLimit?: number };
	repetition: Repetition;
}): SearchContext {
	const descend = createDescend();
	const drawn = createDrawTest(repetition);

	const leaf = createLeaf({
		weights,
		quiescence: options.quiescence ?? false,
		nodeLimit: options.nodeLimit ?? INFINITY,
		descend,
		drawn,
		drawScore: DRAW_SCORE,
	});

	return { descend, repetition, drawn, leaf };
}

// Negamax with alpha-beta, fail-soft. The context travels in the frame rather than in a closure
// so the recursion is a plain function of what it was handed — the frame is allocated per node
// either way, and one more field on it costs nothing next to the closure a factory would need.
//
// The draw test comes before the move list and before the evaluation: the game does not go on
// from a drawn position, so there is nothing to search and nothing to score.
//
// At a leaf the move list is never walked, so it is not built — a mated leaf needs no special
// case, because the evaluation recognises the mate itself and quiescence finds no captures to try
// from one.
export function negamax(frame: Frame): number {
	const { context, position, depth, ply, alpha, beta } = frame;
	const { descend, repetition, leaf } = context;

	if (context.drawn(position)) return DRAW_SCORE;
	if (depth <= 0 || leaf.exhausted()) return leaf.score(frame);

	const moves = legalMoves(position);
	// No moves means the game is over here — mate or stalemate. Hand the check state to the
	// evaluation so `terminalScore` runs one of the two probes rather than both.
	if (moves.length === 0)
		return leaf.evaluate({ position, played: frame.played, ply, inCheck: position.isCheck() });

	let best = -INFINITY;

	// One push for the whole loop: this position is an ancestor of every move tried below it, and
	// it comes off however the loop ends.
	repetition.push(position);

	for (const move of orderMoves({ position, moves })) {
		const score = -negamax({
			context,
			position: descend({ position, move, ply: ply + 1 }),
			played: { parent: position, move },
			depth: depth - 1,
			ply: ply + 1,
			alpha: -beta,
			beta: -Math.max(alpha, best),
		});

		if (score > best) best = score;
		if (best >= beta) break;
	}

	repetition.pop();

	return best;
}

File: engine/src/search/negamax.c
#include <stdbool.h>
#include <stddef.h>

#include "bitboard.h"
#include "draw.h"
#include "eval.h"
#include "evaluate.h"
#include "move.h"
#include "movegen.h"
#include "node.h"
#include "picker.h"
#include "position.h"
#include "search.h"

double no_moves_score(const Search *search, const Position *pos, Node node, bool in_check) {
	if (!in_check) {
		return 0;
	}
	// A bot with no opinion on mate — `givesMate` at zero — evaluates the mated position like any
	// other, so the terminal score falls back on the evaluation rather than on a flat number.
	double score = 0;
	if (terminal_score(pos, search->eval.weights, node.ply, &score)) {
		return score;
	}
	return evaluate_features(&search->eval, pos, node.played);
}

double search_child(Search *search, Position *pos, Node child, int depth, double alpha, double beta,
                    bool first) {
	if (first) {
		return -search_node(search, pos, child, depth, -beta, -alpha);
	}
	double score = -search_node(search, pos, child, depth, -next_up(alpha), -alpha);
	if (score > alpha && score < beta && !search->aborted) {
		score = -search_node(search, pos, child, depth, -beta, -alpha);
	}
	return score;
}

// The draw test comes before everything else: the game does not go on from a drawn position, so
// there is nothing to search and nothing to score. The root never asks it — its position is the
// one on the board, which the game has already let stand.
double search_node(Search *search, Position *pos, Node node, int depth, double alpha, double beta) {
	if (position_drawn(pos, search->history)) {
		return 0;
	}
	if (depth <= 0) {
		return quiesce(search, pos, node, alpha, beta);
	}
	if (exhausted(search)) {
		return 0;
	}
	search->nodes++;
	Picker picker;
	picker_start(&picker, search->table != NULL ? table_move(search->table, pos->hash) : MOVE_NONE,
	             node.ply);
	double best = -__builtin_inf();
	Move best_move = MOVE_NONE;
	history_push(search->history, pos->hash);
	for (Move move; (move = picker_next(&picker, search, pos)) != MOVE_NONE;) {
		Played played = played_move(pos, move);
		Undo undo;
		position_make(pos, move, &undo);
		Node child = {.played = &played, .ply = node.ply + 1};
		double floor = alpha > best ? alpha : best;
		double score =
		    search_child(search, pos, child, depth - 1, floor, beta, best_move == MOVE_NONE);
		position_unmake(pos, move, &undo);
		// An aborted child's score is whatever the search was holding when it stopped: nothing
		// is learnt from it, not a cutoff and not a table entry.
		if (search->aborted) {
			history_pop(search->history);
			return 0;
		}
		if (score > best || best_move == MOVE_NONE) {
			best = score;
			best_move = move;
		}
		if (best >= beta) {
			if (!is_capture(pos, move) && move_promotion(move) == PAWN) {
				record_cutoff(search, pos, move, node.ply, depth);
			}
			break;
		}
	}
	history_pop(search->history);
	if (best_move == MOVE_NONE) {
		return no_moves_score(search, pos, node, move_context(pos).checkers != 0);
	}
	if (search->table != NULL) {
		table_store(search->table, pos->hash, best_move);
	}
	return best;
}
